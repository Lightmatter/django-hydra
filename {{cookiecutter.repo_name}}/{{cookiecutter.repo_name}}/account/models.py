from django.core.mail import send_mail
from django.db import models

from django.contrib.auth.models import (
    AbstractBaseUser,
    BaseUserManager,
    PermissionsMixin,
)
from django.utils.translation import ugettext_lazy as _
from model_utils.models import TimeStampedModel


class UserManager(BaseUserManager):
    use_in_migrations = True

    @classmethod
    def normalize_email(cls, email):
        """All email providers treat emails in a case-insensitive manner."""
        email = email or ""
        return email.lower()

    def _create_user(self, email, password, is_staff, is_superuser, **extra_fields):
        email = self.normalize_email(email)
        user = self.model(
            email=email,
            is_staff=is_staff,
            is_active=True,
            is_superuser=is_superuser,
            **extra_fields
        )
        user.set_password(password)
        user.save(using=self._db)
        return user

    def create_user(self, email, password=None, **extra_fields):
        return self._create_user(email, password, False, False, **extra_fields)

    def create_superuser(self, email, password, **extra_fields):
        return self._create_user(email, password, True, True, **extra_fields)

    def get_by_natural_key(
        self, username
    ):  # Used by contrib.auth.backends.ModelBackend
        return self.get(email__iexact=username)


class User(TimeStampedModel, AbstractBaseUser, PermissionsMixin):
    email = models.EmailField(
        _("email address"),
        unique=True,
        error_messages={"unique": _("A user with that email address already exists.")},
    )
    first_name = models.CharField(_("first name"), max_length=30, blank=True)
    last_name = models.CharField(_("last name"), max_length=30, blank=True)
    is_staff = models.BooleanField(
        _("staff status"),
        default=False,
        help_text=_("Designates whether the user can log into this admin " "site."),
    )
    is_active = models.BooleanField(
        _("active"),
        default=True,
        help_text=_(
            "Designates whether this user should be treated as "
            "active. Unselect this instead of deleting accounts."
        ),
    )

    objects = UserManager()

    USERNAME_FIELD = "email"
    REQUIRED_FIELDS = ["first_name", "last_name"]

    class Meta:
        verbose_name = _("user")
        verbose_name_plural = _("users")

    def get_full_name(self):
        """Returns the first_name plus the last_name, with a space in between."""
        full_name = "%s %s" % (self.first_name, self.last_name)
        return full_name.strip()

    def get_short_name(self):
        """Returns the short name for the user."""
        return self.first_name

    def email_user(self, subject, message, from_email=None, **kwargs):
        """Sends an email to this User."""
        send_mail(subject, message, from_email, [self.email], **kwargs)

    def __str__(self):
        name = self.get_full_name()
        return name if name else self.email

    def get_absolute_url(self):
        """
        The absolute url of the user model
        """
        raise NotImplementedError()
