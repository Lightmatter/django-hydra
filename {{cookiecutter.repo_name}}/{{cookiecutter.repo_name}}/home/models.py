from django.db import models  # NOQA

from django_extensions.db.models import AutoSlugField  # NOQA
from model_utils.models import TimeStampedModel  # NOQA

# Create your models here.
