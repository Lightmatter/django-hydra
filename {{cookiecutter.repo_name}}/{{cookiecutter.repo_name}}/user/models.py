from django.contrib import auth
from django.contrib.auth.hashers import make_password
from django.contrib.auth.models import AbstractBaseUser, BaseUserManager, PermissionsMixin
from django.core.mail import send_mail
from django.db import models
from django.utils.translation import gettext_lazy as _
from model_utils.models import TimeStampedModel


class UserManager(BaseUserManager):
    use_in_migrations = True

    @classmethod
    def normalize_email(cls, email):
        """All email providers treat emails in a case-insensitive manner."""
        email = email or ""
        return email.lower()

    def _create_user(self, email, password, **extra_fields):
        email = self.normalize_email(email)
        user = self.model(
            email=email,
            is_active=True,
            **extra_fields,
        )
        user.password = make_password(password)
        user.set_password(password)
        user.save(using=self._db)
        return user

    def create_user(self, email=None, password=None, **extra_fields):
        extra_fields.setdefault("is_staff", False)
        extra_fields.setdefault("is_superuser", False)
        return self._create_user(email, password, **extra_fields)

    def create_superuser(self, email=None, password=None, **extra_fields):
        extra_fields.setdefault("is_staff", True)
        extra_fields.setdefault("is_superuser", True)

        if extra_fields.get("is_staff") is not True:
            raise ValueError("Superuser must have is_staff=True.")
        if extra_fields.get("is_superuser") is not True:
            raise ValueError("Superuser must have is_superuser=True.")

        return self._create_user(email, password, **extra_fields)

    def get_by_natural_key(self, username):  # Used by contrib.auth.backends.ModelBackend
        return self.get(email__iexact=username)

    def with_perm(  # noqa
        self,
        perm,
        is_active=True,
        include_superusers=True,
        backend=None,
        obj=None,
    ):
        if backend is None:
            backends = auth._get_backends(return_tuples=True)  # noqa
            if len(backends) == 1:
                backend, _ = backends[0]
            else:
                raise ValueError(
                    "You have multiple authentication backends configured and "
                    "therefore must provide the `backend` argument."
                )
        elif not isinstance(backend, str):
            raise TypeError(f"backend must be a dotted import path string (got {backend}).")
        else:
            backend = auth.load_backend(backend)
        if hasattr(backend, "with_perm"):
            return backend.with_perm(
                perm,
                is_active=is_active,
                include_superusers=include_superusers,
                obj=obj,
            )
        return self.none()


class User(TimeStampedModel, AbstractBaseUser, PermissionsMixin):
    email = models.EmailField(
        _("email address"),
        unique=True,
        error_messages={"unique": _("A user with that email address already exists.")},
    )
    first_name = models.CharField(_("first name"), max_length=30, blank=True)
    last_name = models.CharField(_("last name"), max_length=30, blank=True)
    is_staff = models.BooleanField(
        _("staff status"),
        default=False,
        help_text=_("Designates whether the user can log into this admin site."),
    )
    is_active = models.BooleanField(
        _("active"),
        default=True,
        help_text=_(
            "Designates whether this user should be treated as "
            "active. Unselect this instead of deleting accounts."
        ),
    )

    objects = UserManager()

    EMAIL_FIELD = "email"
    USERNAME_FIELD = "email"
    REQUIRED_FIELDS = ["first_name", "last_name"]

    class Meta:
        verbose_name = _("user")
        verbose_name_plural = _("users")

    def clean(self):
        super().clean()
        self.email = self.__class__.objects.normalize_email(self.email)

    def get_full_name(self):
        """Returns the first_name plus the last_name, with a space in between."""
        full_name = f"{self.first_name} {self.last_name}"
        return full_name.strip()

    def get_short_name(self):
        """Returns the short name for the user."""
        return self.first_name

    def email_user(self, subject, message, from_email=None, **kwargs):
        """Sends an email to this User."""
        send_mail(subject, message, from_email, [self.email], **kwargs)

    def __str__(self):
        name = self.get_full_name()
        return name if name else self.email

    def get_absolute_url(self):
        """
        The absolute url of the user model
        """
        raise NotImplementedError()
