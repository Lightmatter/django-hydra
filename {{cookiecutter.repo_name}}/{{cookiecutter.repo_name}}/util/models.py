from django.db import models

from django.utils.translation import ugettext_lazy as _

from .util import file_url


class TestFileModel(models.Model):
    file_field = models.ImageField(_("foo"), upload_to=file_url("filez"))
