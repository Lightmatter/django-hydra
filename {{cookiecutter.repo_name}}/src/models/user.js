import React, { useCallback } from 'react';
import axios from 'util/axios';
import { EMAIL, GENERIC_FIELD_ERROR, REQUIRED, TOO_LONG, TOO_SHORT } from '../constants.js';
import * as Yup from 'yup';
import constate from 'constate';

import useSWR from 'swr';

function equalTo(ref, msg) {
  ref = Yup.ref(ref);
  return this.test({
    name: 'equalTo',
    exclusive: false,
    message: msg || `${ref.path} must be the same!`,
    params: {
      reference: ref.path,
    },
    test: function(value) {
      return value === this.resolve(ref);
    },
  });
}

Yup.addMethod(Yup.string, 'equalTo', equalTo);

export const USER_ME = 'http://localhost:8000/auth/users/me/';

const AuthSchema = {
  first_name: Yup.string()
    .min(2, TOO_SHORT)
    .max(50, TOO_LONG)
    .required(REQUIRED),
  last_name: Yup.string()
    .min(2, TOO_SHORT)
    .max(50, TOO_LONG)
    .required(REQUIRED),
};

export const SignupSchema = Yup.object().shape({
  ...AuthSchema,
  email: Yup.string()
    .email(EMAIL)
    .required(REQUIRED),
  password: Yup.string()
    .required(REQUIRED)
    .min(6, TOO_SHORT),
  re_password: Yup.string()
    .required(REQUIRED)
    .equalTo('password', 'The Two Passwords Must Match'),
});

export const LoginSchema = Yup.object().shape({
  password: Yup.string()
    .required(REQUIRED)
    .min(6, TOO_SHORT),
  email: Yup.string()
    .email(EMAIL)
    .required(REQUIRED),
});

function handleApiErrors(error) {
  let err;
  if (error.response) {
    /*
     * The request was made and the server responded with a
     * status code that falls out of the range of 2xx
     */
    err = error.response.data;
  } else if (error.request || error.isAxiosError) {
    /*
     * The request was made but no response was received, `error.request`
     * is an instance of XMLHttpRequest in the browser and an instance
     * of http.ClientRequest in Node.js
     */
    err = { non_field_errors: 'There was a problem processing your request.' };
  } else {
    // Something happened in setting up the request and triggered an Error
    err = error;
  }
  return Promise.reject(err);
}

export function getCurrentUserDetails() {
  return axios.get(USER_ME); //todo  - create hook to manage current user and set global state
}

export function registerUser(userData) {
  const url = `http://localhost:8000/auth/users/`;

  return axios
    .post(url, userData)
    .then(response => {})
    .catch(error => {
      return handleApiErrors(error);
    });
}

export function updateUser(userData, userId) {
  const url = USER_ME;
  //todo  - create hook to manage current user and set global state
  return axios.put(url, userData).catch(error => {
    return handleApiErrors(error);
  });
}

export function logIn(userData) {
  const url = `http://localhost:8000/auth/token/login/`;
  return axios
    .post(url, userData, {
      //AxiosRequestConfig parameter
      withCredentials: true, //correct
    })
    .then(response => {
      // use login session, so this should set a cookie but return a token. We still love you token.
      const token = `Token ${response.data['key']}`;
    })
    .catch(error => {
      return handleApiErrors(error);
    });
}

export function logOut() {
  const url = `/auth/token/logout/`;
  return axios
    .post(url)
    .then(() => {})
    .then(() => {}) //todo: clear data, use local session event hook to notify other tabs
    .catch(error => {
      return handleApiErrors(error);
    });
}

export function forgotPass(email) {
  const url = `/auth/users/reset_password/`;
  return axios.post(url, email).catch(error => {
    return handleApiErrors(error);
  });
}

export function confirmForgotPass(userInfo) {
  const url = `/auth/users/reset_password_confirm/`;
  return axios.post(url, userInfo).catch(error => {
    return handleApiErrors(error);
  });
}

export function confirmEmail(uid, token) {
  const url = `/auth/users/activation/`;
  const data = { uid, token };
  return axios.post(url, data).catch(error => {
    return handleApiErrors(error);
  });
}

export function resendConfirmEmail(email) {
  const url = '/auth/users/resend_activation';
  const data = { email };
  return axios.post(url).catch(error => {
    return handleApiErrors(error);
  });
}

export function useCurrentUserSWR() {
  const options = {
    shouldRetryOnError: false,
    onSuccess: function(data, key, config) {},
    onError: function(err, key, config) {},
  };
  const { data, error, isValidating, mutate } = useSWR(
    USER_ME,
    query =>
      axios({
        method: 'get',
        url: query,
        headers: { Accept: '*/*' }, // as fetch for preload doesn't set accept correctly
      }),
    options
  );
  const isAuthenticated = Boolean(data);
  const user = isAuthenticated ? data.data.data.attributes : {};
  return { user, isAuthenticated, error, isValidating, mutate };
}

export const [
  CurrentUserProvider,
  useCurrentUser,
  useIsAuthenticated,
  useCurrentUserError,
  useCurrentUserIsValidating,
  useMutateCurrentUser,
] = constate(
  useCurrentUserSWR,
  value => value.user,
  value => value.isAuthenticated,
  value => value.error,
  value => value.useCurrentUserIsValidating,
  value => useMutateCurrentUser
);
