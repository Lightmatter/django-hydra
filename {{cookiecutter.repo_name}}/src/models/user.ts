import { useEffect, useState } from 'react';
import * as Yup from 'yup';
import constate from 'constate';
import { useSnackbar } from 'notistack';
import useSWR from 'swr';

import axios from 'util/axios';
import { EMAIL, REQUIRED, TOO_LONG, TOO_SHORT } from '../constants';

export const USER_ME = '/auth/users/me/';

function equalTo(yupRef, msg) {
  const ref = Yup.ref(yupRef);
  return this.test({
    name: 'equalTo',
    exclusive: false,
    message: msg || `${ref.path} must be the same!`,
    params: {
      reference: ref.path,
    },
    test(value) {
      return value === this.resolve(ref);
    },
  });
}

Yup.addMethod(Yup.string, 'equalTo', equalTo);
const name = Yup.string()
  .min(2, TOO_SHORT)
  .max(50, TOO_LONG)
  .required(REQUIRED);
const password = Yup.string().required(REQUIRED).min(6, TOO_SHORT);
const email = Yup.string().email(EMAIL).required(REQUIRED);
const UserDetailSchema = {
  firstName: name,
  lastName: name,
};

const SetPassSchema = {
  password,
  rePassword: password.equalTo('password', 'The Two Passwords Must Match'),
};

export const ProfileSchema = Yup.object().shape({
  ...UserDetailSchema,
});
export const SignupSchema = Yup.object().shape({
  ...UserDetailSchema,
  ...SetPassSchema,
  email,
  tos: Yup.boolean().required(REQUIRED).oneOf([true], 'Field must be checked'),
});

export const LoginSchema = Yup.object().shape({
  password,
  email,
});

export const ForgotPassSchema = Yup.object().shape({
  email,
});

export const ResetPassSchema = Yup.object().shape({
  newPassword: password,
  reNewPassword: password.equalTo(
    'newPassword',
    'The Two Passwords Must Match'
  ),
});

export const ChangePassSchema = Yup.object().shape({
  currentPassword: password,
  newPassword: password,
  reNewPassword: password.equalTo(
    'newPassword',
    'The Two Passwords Must Match'
  ),
});

export const ChangeEmailSchema = Yup.object().shape({
  newEmail: email,
  reNewEmail: email.equalTo('newEmail', 'The Two Emails must match'),
  currentPassword: password,
});
export const DeleteUserSchema = Yup.object().shape({
  currentPassword: password,
});

export function registerUser(userData) {
  const url = '/auth/register/';

  return axios.post(url, userData);
}

export function updateUser(userData) {
  return axios.put(USER_ME, userData);
}

export function deleteUser(data) {
  const url = '/auth/users/me/';
  return axios.delete(url, { data }).then(() => {
    window.dispatchEvent(new Event('logout'));
    window.localStorage.setItem('logout', Date.now());
  });
}

export function logIn(userData) {
  const url = '/auth/token/login/';
  return axios.post(url, userData).then(() => {
    // use login session, so this should set a cookie but return a token. We still love you token.
    /* const token = `Token ${response.data.key}`; */
    // notify app that we've logged in
    window.dispatchEvent(new Event('login'));
    window.localStorage.setItem('login', Date.now());
  });
}

export function logOut() {
  const url = '/auth/token/logout/';
  return axios.post(url).then(() => {
    window.dispatchEvent(new Event('logout'));
    window.localStorage.setItem('logout', Date.now());
    window.location = '/';
  });
}

export function forgotPass(userEmail) {
  const url = '/auth/users/reset_password/';
  return axios.post(url, userEmail);
}

export function resetPass(userInfo) {
  const url = '/auth/users/reset_password_confirm/';
  return axios.post(url, userInfo);
}

export function changePass(data) {
  const url = '/auth/users/set_password/';
  return axios.post(url, data);
}

export function changeEmail(data) {
  const url = '/auth/users/set_email/';
  return axios.post(url, data);
}

export function confirmEmail(uid, token) {
  const url = '/auth/users/activation/';
  const data = { uid, token };
  return axios.post(url, data);
}

export function resendConfirmEmail() {
  const url = '/auth/users/resend_activation';
  return axios.post(url);
}
let isAuthenticated;
let setAuthenticated;

export function useCurrentUserSWR({ initialUser }) {
  [isAuthenticated, setAuthenticated] = useState(Boolean(initialUser));
  const { enqueueSnackbar } = useSnackbar();
  const options = {
    shouldRetryOnError: false,
    onSuccess: () => {
      if (isAuthenticated === false) {
        setAuthenticated(true);
      }
    },
    onError: err => {
      if (
        err.isAxiosError &&
        err.response?.status === 403 &&
        isAuthenticated === true
      ) {
        setAuthenticated(false);
        enqueueSnackbar('Something caused you to be logged out!', {
          variant: 'warning',
        });
      }
    },
    // revalidateOnFocus: isAuthenticated, //TODO: Currently a bug in useSWR - this doesn't change between renders
    initialData: initialUser,
  };

  // TODO: this will make a request to the server on tab focus if you're logged out.
  // really we shouldn't do that - we know if the revalidation attempt will be successful or not based on isauthenticated
  // This could be smarter w/ a stateful representation of isAuthenticated, vs a derived one to check to test.
  const { data, error, isValidating, mutate } = useSWR(
    USER_ME,
    query =>
      axios({
        method: 'get',
        url: query,
        headers: { Accept: '*/*' }, // as fetch for preload doesn't set accept correctly
      }).then(response => response.data),
    options
  );
  const user = isAuthenticated ? data : null;

  useEffect(() => {
    const messageOtherTab = event => {
      if (event.key === 'logout') {
        enqueueSnackbar('You logged out in another tab!', {
          variant: 'warning',
        });
      } else if (event.key === 'login') {
        enqueueSnackbar('You logged in on another tab!', {
          variant: 'info',
        });
      }
    };

    const messageThisTab = event => {
      if (event.type === 'logout') {
        enqueueSnackbar('Successfully logged out', {
          variant: 'success',
        });
      } else if (event.type === 'login') {
        enqueueSnackbar('Successfully logged in', {
          variant: 'success',
        });
      }
    };

    const syncLogout = event => {
      if (event.key === 'logout' || event.type === 'logout') {
        setAuthenticated(false);
        mutate(null, false);
      }
    };
    const syncLogin = event => {
      if (event.key === 'login' || event.type === 'login') {
        setAuthenticated(true);
        mutate();
      }
    };

    window.addEventListener('storage', messageOtherTab);
    window.addEventListener('login', messageThisTab);
    window.addEventListener('logout', messageThisTab);

    window.addEventListener('storage', syncLogout);
    window.addEventListener('logout', syncLogout);

    window.addEventListener('storage', syncLogin);
    window.addEventListener('login', syncLogin);

    return () => {
      window.removeEventListener('storage', syncLogout);
      window.removeEventListener('logout', syncLogout);
      window.localStorage.removeItem('logout');

      window.removeEventListener('storage', syncLogin);
      window.removeEventListener('login', syncLogin);
      window.localStorage.removeItem('login');

      window.removeEventListener('storage', messageOtherTab);
      window.removeEventListener('login', messageThisTab);
      window.removeEventListener('logout', messageThisTab);
    };
  }, []);

  return { user, isAuthenticated, error, isValidating, mutate };
}

export const [
  CurrentUserProvider,
  useCurrentUser,
  useIsAuthenticated,
  useCurrentUserError,
  useCurrentUserIsValidating,
  useMutateCurrentUser,
] = constate(
  useCurrentUserSWR,
  value => value.user,
  value => value.isAuthenticated,
  value => value.error,
  value => value.isValidating,
  value => value.mutate
);
