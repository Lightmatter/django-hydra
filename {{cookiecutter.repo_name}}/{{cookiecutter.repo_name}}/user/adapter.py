from allauth.account.adapter import DefaultAccountAdapter
from django.http import HttpResponseRedirect
from django.urls import reverse
from django_htmx.http import HttpResponseClientRedirect


class HTMXAccountAdapter(DefaultAccountAdapter):
    def respond_email_verification_sent(self, request, user):
        url = reverse("account_email_verification_sent")
        if request.htmx:
            return HttpResponseClientRedirect(url)
        return HttpResponseRedirect(url)
