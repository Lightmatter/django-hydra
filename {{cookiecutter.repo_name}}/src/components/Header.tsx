import {
  usePopupState,
  bindTrigger,
  bindMenu,
} from 'material-ui-popup-state/hooks';
import { useRef } from 'react';

import {
  AppBar,
  Toolbar,
  Typography,
  IconButton,
  Menu,
  MenuItem,
  Grid,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import MenuIcon from '@material-ui/icons/Menu';
import AccountCircle from '@material-ui/icons/AccountCircle';
import Router from 'next/router';

import Link from 'components/router/Link';
import { useCurrentUser, logOut } from 'models/user';

const useStyles = makeStyles(theme => ({
  menuButton: {
    marginRight: theme.spacing(2),
  },
  title: {
    flexGrow: 1,
  },
}));

export default function MenuAppBar() {
  const classes = useStyles();
  const anchorRef = useRef(null);
  const navMenuState = usePopupState({
    variant: 'popover',
    popupId: 'navMenu',
  });
  const profileMenuState = usePopupState({
    variant: 'popover',
    popupId: 'profileMenu',
  });
  const user = useCurrentUser();
  const logoutButtonClick = () => {
    logOut();
    profileMenuState.close();
  };

  const handleMenuCloseWithLink = ({ event, link = '', callback = null }) => {
    if (anchorRef.current && anchorRef.current.contains(event.target)) {
      return;
    }

    if (link) {
      Router.push(link);
    }

    if (callback) {
      callback();
    }

    profileMenuState.close();
    navMenuState.close();
  };
  return (
    <AppBar position="static">
      <Toolbar>
        <IconButton
          edge="start"
          className={classes.menuButton}
          color="inherit"
          aria-label="menu"
          data-cy="auth-menu-button"
          {...bindTrigger(navMenuState)}
        >
          <MenuIcon />
        </IconButton>
        <Grid container justifyContent="space-between" alignItems="center">
          <Grid item xs={6} md={3}>
            <Menu keepMounted {...bindMenu(navMenuState)}>
              <MenuItem
                data-cy="login"
                onClick={event =>
                  handleMenuCloseWithLink({
                    event,
                    link: '/login',
                  })
                }
              >
                Login
              </MenuItem>
              <MenuItem
                data-cy="signup"
                onClick={event =>
                  handleMenuCloseWithLink({
                    event,
                    link: '/signup',
                  })
                }
              >
                Sign Up
              </MenuItem>
              <MenuItem
                onClick={event =>
                  handleMenuCloseWithLink({
                    event,
                    link: '/error-page',
                  })
                }
              >
                Throw an error
              </MenuItem>
            </Menu>
            <Link href="/">
              <Typography
                variant="h6"
                color="textPrimary"
                className={classes.title}
              >
                Lightmatter
              </Typography>
            </Link>
          </Grid>
          <Grid item xs={6} md={9} align="right">
            {user && (
              <div>
                <span data-cy="logged-in-name">Hey {user.firstName}</span>
                <IconButton
                  aria-label="account of current user"
                  aria-controls="menu-appbar"
                  aria-haspopup="true"
                  color="inherit"
                  {...bindTrigger(profileMenuState)}
                >
                  <AccountCircle />
                </IconButton>
                <Menu id="menu-appbar" {...bindMenu(profileMenuState)}>
                  <MenuItem
                    onClick={event =>
                      handleMenuCloseWithLink({
                        event,
                        link: '/account',
                      })
                    }
                  >
                    Profile
                  </MenuItem>
                  <MenuItem
                    onClick={event =>
                      handleMenuCloseWithLink({
                        event,
                        link: '/account/change-password',
                      })
                    }
                  >
                    Change Password
                  </MenuItem>
                  <MenuItem
                    onClick={event =>
                      handleMenuCloseWithLink({
                        event,
                        link: '/account/change-email',
                      })
                    }
                  >
                    Change Email
                  </MenuItem>
                  <MenuItem
                    onClick={event =>
                      handleMenuCloseWithLink({
                        event,
                        link: '/account/delete-account',
                      })
                    }
                  >
                    Delete Account
                  </MenuItem>
                  <MenuItem
                    onClick={event =>
                      handleMenuCloseWithLink({
                        event,
                        callback: logoutButtonClick,
                      })
                    }
                  >
                    Logout
                  </MenuItem>
                </Menu>
              </div>
            )}
          </Grid>
        </Grid>
      </Toolbar>
    </AppBar>
  );
}
