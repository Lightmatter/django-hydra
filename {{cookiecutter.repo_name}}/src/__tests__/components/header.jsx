import React from 'react';
import { createMount } from '@material-ui/core/test-utils';

import renderer from 'react-test-renderer';
import Header from 'components/Header';

import { SnackbarProvider } from 'notistack';
import { CurrentUserProvider } from 'models/user';
import { act } from 'react-dom/test-utils';

describe('Header', () => {
  it('renders header unchanged', () => {
    const renderedValue = createMount()(
      <SnackbarProvider>
        <CurrentUserProvider>
          <Header />
        </CurrentUserProvider>
      </SnackbarProvider>
    );

    expect(renderedValue.html()).toMatchSnapshot();
  });
  it('renders displays the users first name when booting  login', async () => {
    const user = { firstName: 'Ben' };
    const renderedValue = createMount()(
      <SnackbarProvider>
        <CurrentUserProvider initialUser={user}>
          <Header />
        </CurrentUserProvider>
      </SnackbarProvider>
    );
    expect(renderedValue.html()).toMatch('Hey Ben');
  });
  it('renders nothing for username without user ', async () => {
    const renderedValue = createMount()(
      <SnackbarProvider>
        <CurrentUserProvider initialUser={null}>
          <Header />
        </CurrentUserProvider>
      </SnackbarProvider>
    );
    expect(renderedValue.html()).not.toMatch('Hey');
  });
  it('renders a username after logging in ', async () => {});
  it('stops rendering a username after logging out ', async () => {});
});
