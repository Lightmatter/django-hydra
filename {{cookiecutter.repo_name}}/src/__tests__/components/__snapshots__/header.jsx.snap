// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders header unchanged 1`] = `"<header class=\\"MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4\\"><div class=\\"MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters\\"><button class=\\"MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-1 MuiIconButton-colorInherit MuiIconButton-edgeStart\\" tabindex=\\"0\\" type=\\"button\\" aria-label=\\"menu\\" data-cy=\\"auth-menu-button\\" aria-haspopup=\\"true\\"><span class=\\"MuiIconButton-label\\"><svg class=\\"MuiSvgIcon-root\\" focusable=\\"false\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\"><path d=\\"M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z\\"></path></svg></span><span class=\\"MuiTouchRipple-root\\"></span></button><div class=\\"MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between\\"><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-3\\"><a class=\\"MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary\\" href=\\"/\\"><h6 class=\\"MuiTypography-root makeStyles-title-2 MuiTypography-h6 MuiTypography-colorTextPrimary\\">Lightmatter</h6></a></div><div class=\\"MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-9\\" align=\\"right\\"></div></div></div></header>"`;
